File: By-Swift/By-Swift/Leetcode/course-schedule.swift
class Solution {
    func dfs(_ idx: Int, _ graph: [[Int]], _ visited: inout [String]) -> Bool {
        if visited[idx] == "doing" {
            return false
        }
        if visited[idx] == "done" {
            return true
        }
        visited[idx] = "doing"
        for g in graph[idx] {
            if !dfs(g, graph, &visited) {
                return false
            }
        }
        visited[idx] = "done"        
        return true
    }
    
    func canFinish(_ numCourses: Int, _ prerequisites: [[Int]]) -> Bool {
        var graph = Array(repeating: [Int](), count: numCourses)
        var visited = Array(repeating: "notYet", count: numCourses)
        
        for course in prerequisites {
            graph[course.first!].append(course.last!)   
        }
        
        for idx in 0..<numCourses {
            if !dfs(idx, graph, &visited) {
                return false
            }
        }        
        
        return true
    }
}

File: By-Python/leetcode/Course-Schedule.py
class Solution:
    def dfs(self, idx, visited, graph):
        if visited[idx] == "doing":
            return False
        if visited[idx] == "done":
            return True
        
        visited[idx] = "doing"
        
        for g in graph[idx]:
            if not self.dfs(g, visited, graph):
                return False   
            
        visited[idx] = "done"
        return True
    
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        graph = [[] for _ in range(numCourses)]
        visited = ["notYet" for _ in range(numCourses)] 
        
        for x,y in prerequisites:
            graph[x].append(y)
        
        for i in range(numCourses):
            if not self.dfs(i, visited, graph):
                return False
            
        return True